### PR TITLE
Added validate method to Application server

### PR DESCRIPTION
**Pull request by josecordaz**
Closes: #644 

**What changed?**
Added end point Validate to generate JTW tokens

**Why?**
This was added as a first stage to start securing the access to the api

**How did you test it?**
Added unit tests

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: api/applications/applications.proto
Comment (jpellizzari) [inline api/applications/applications.proto:24]: Let re-work this comment. It doesn't really do this in our case. This will end up as auto-generated API documentation.

File: api/applications/applications.proto
Comment (jpellizzari) [inline api/applications/applications.proto:79]: Need comments here, and lets try to get these aligned.

File: api/applications/applications.proto
Comment (jpellizzari) [inline api/applications/applications.proto:83]: Comment needs a re-work

File: pkg/server/server.go
Comment (jpellizzari) [inline pkg/server/server.go:322]: Let's use `NormalizedRepoURL` here

File: pkg/server/server.go
Comment (jpellizzari) [inline pkg/server/server.go:320]: Nit: the convention we have been using is to label the request payload `msg`. `r` is generally reserved for an `http.Request` object.

File: pkg/services/auth/jwt.go
Comment (jpellizzari) [inline pkg/services/auth/jwt.go:24]: We have a typed provider name now. I believe its `gitproviders.GitProviderNameGithub`

File: pkg/services/auth/jwt_test.go
Comment (jpellizzari) [inline pkg/services/auth/jwt_test.go:28]: Let's avoid checking a specific error message. Can we check for a specific error `type` instead?

File: pkg/server/server_test.go
Comment (jpellizzari) [inline pkg/server/server_test.go:110]: Seems like an odd assertion here. We are asserting that an error is NOT like this string? So it could be anything. Custom error types will make this safer.

File: pkg/server/server_test.go
Comment (jpellizzari) [inline pkg/server/server_test.go:106]: We should be testing for a specific HTTP error code here

**Review by jpellizzari — COMMENTED** [review]
Looking good. I listed some changes that I think will really improve maintainability.

File: pkg/server/server.go
Comment (josecordaz) [reply pkg/server/server.go:322]: not sure I understand this 🤔, what I'm trying to do is just to make sure the provider name passed in is either `github` or `gitlab` using the constants from `go-git-providers` 🙂

File: pkg/services/auth/jwt_test.go
Comment (josecordaz) [reply pkg/services/auth/jwt_test.go:28]: that sounds like a good idea, but not really sure how to do it when the error is dynamic 🤔 . That is why I'm using RegExp here. It could fail with `invalid token: token is expired by 4s` or `invalid token: token is expired by 3s`.

File: pkg/server/server_test.go
Comment (josecordaz) [reply pkg/server/server_test.go:110]: good catch 🙂

File: pkg/server/server_test.go
Comment (josecordaz) [reply pkg/server/server_test.go:106]: not sure I understand, this is one of the possible errors `rpc error: code = Unknown desc = unknown provider name wrongprovider, expecting github or gitlab`

Comment (jpellizzari) [reply pkg/server/server_test.go:106]: Sorry I thought we had a reference to HTTP errors already. 

You can use the runtime package to return a specific error code. An error case might look like this:
```
	if err != nil {
		return nil, &runtime.HTTPStatusError{HTTPStatus: http.StatusUnauthorized, Err: err}
	}
```

We need to ensure the correct error code is being returned. The UI and API users will configure their behavior based on this code.

File: pkg/services/auth/jwt_test.go
Comment (jpellizzari) [reply pkg/services/auth/jwt_test.go:29]: There is no need to include the token expiration delta in the message. That might actually be a security risk.

File: pkg/server/server.go
Comment (jpellizzari) [reply pkg/server/server.go]: `NormalizedRepoURL` has some helper methods to return a type-safe git provider name: https://github.com/weaveworks/weave-gitops/blob/main/pkg/gitproviders/provider.go#L624

Comment (jpellizzari) [reply pkg/server/server.go]: Nevermind I misunderstood. Let's try using the typesafe provider names rather than looking at string prefixes

File: pkg/services/auth/jwt_test.go
Comment (josecordaz) [reply pkg/services/auth/jwt_test.go:29]: I'll replace the error msg coming from the jwt library with `unauthorized token`, not sure if that would be accurate 🤔

Comment (jpellizzari) [reply pkg/services/auth/jwt_test.go:29]: Testing for another library's error string is a sure fire way to have flaky tests. If the other lib changes the error message by 1 character, unit tests will suddenly break once we upgrade that dependency.

Comment (josecordaz) [reply pkg/services/auth/jwt_test.go]: @jpellizzari I have receive feedback from James to assert error messages in previous PRs instead of just doing `.HaveOccurred()`. But I could be misunderstanding something. I'm not sure what I have to do here. @JamWils would you mind to help me clear this up ?

File: pkg/services/auth/jwt.go
Comment (luizbafilho) [inline pkg/services/auth/jwt.go:27]: I think `Generate` it's a bit too generic, I think `GenerateJWT` is probably easier to read and  disambiguate from the provider token

File: pkg/services/auth/jwt.go
Comment (luizbafilho) [inline pkg/services/auth/jwt.go:41]: Same as `Generate`

File: pkg/services/auth/jwt.go
Comment (luizbafilho) [inline pkg/services/auth/jwt.go:55]: Instead of checking error strings, we could use their error types and have a more robust error checking: https://pkg.go.dev/github.com/golang-jwt/jwt#example-Parse-ErrorChecking

**Review by luizbafilho — CHANGES_REQUESTED** [review]
Looks good, only a few comments and the big one is moving from `github.com/dgrijalva/jwt-go` to `https://github.com/golang-jwt/jwt` the former is no longer maintained and its code was migrated to the latter.

File: pkg/services/auth/jwt.go
Comment (luizbafilho) [inline pkg/services/auth/jwt.go:59]: Looking at this again, if we have an error we'll have an Unauthorized token anyway, what do you think? should we always return that `ErrUnauthorizedToken` and not check what type of error is that since we only care that the valid or not?

So we would become just:

```
if err != nil {
   return nil, ErrUnauthorizedToken
}
```

File: pkg/services/auth/jwt_test.go
Comment (JamWils) [reply pkg/services/auth/jwt_test.go]: Yes, I told @josecordaz to also verify error messages as well.  The example we were looking at when we chatted was you couldn't tell which tests were testing which error path.  By verifying the error messages you actually were reducing flakiness in the test.  I also prefer specificity with assertions.  It looks like we need to have a conversation as a team what the expectations are, pun intended.  Jordan I do agree with you about testing another library's error string.  To eliminate flakiness I would use a pattern such as:

```
func TestSomething(t *T) {
    ....
    err := MyDailyFunc("", "")
    
    expectedErr := json.Unmarshal(badData, &obj)
   t.Assert(err, Equals, fmt.Errorf("MyDailyFunc json unmarshal fail: %w, expectedErr))
}
```

In this instance, I actually propose we verify the string, not for my normal specificity reasons (error messages shouldn't change that often anyway), but for security reasons.  We want to lock-in the contract for the test and if someone changes the error message and puts sensitive information in it that we don't want returned then the test will fail.

We can revisit assertions outside of this PR.  I made my case but leave the final call up to you and Jose.

File: pkg/services/auth/jwt.go
Comment (JamWils) [inline pkg/services/auth/jwt.go]: I had a chat with @josecordaz f2f and we talked about putting `Generate` and `Verify` in an interface so he can mock it out and call it from the `ApplicationServer` more easily.

Comment (jpellizzari) [reply pkg/services/auth/jwt.go]: Why mock it out?

**Comment by luizbafilho** [comment]
on the move to https://github.com/golang-jwt/jwt let's use the latest release v4, here is the migration guide https://github.com/golang-jwt/jwt/blob/main/MIGRATION_GUIDE.md. Quite simple since it's backward compatibly.

File: pkg/services/auth/jwt.go
Comment (josecordaz) [reply pkg/services/auth/jwt.go:59]: Good point.  I'm thinking in the user consuming the api, should the user get a specific error to acknowledge that the api  call was done wrong?. I know we don't want to return `ValidationErrorExpired` because that exposes how long ago the token expired (security reasons) .  I guess this comes down to which errors we consider are security issues. Do all errors that `ParseWithClaims` returns would be considered security issues? @jpellizzari @JamWils any thoughts you might have ?

Comment (jpellizzari) [reply pkg/services/auth/jwt.go:59]: > Do all errors that ParseWithClaims returns would be considered security issues?

Probably? This error happens when someone submits a malformed token, which **is** an `Unauthorized` token. I would be okay with returning an `ErrUnauthorizedToken` here.

File: pkg/services/auth/jwt_test.go
Comment (jpellizzari) [reply pkg/services/auth/jwt_test.go]: I like the solution @josecordaz has now with keeping the assertion in sync with the error string. The important thing is to verify that the the error you are receiving is the one you expect, so we agree there.

I'm trying to to have tests break when trivial error messages changes happen. If as assertion has a string literal, you will have to update the assertion every time the error message changes. Having both the assertion AND the error message refer to the same string is :+1:

**Review by jpellizzari — APPROVED** [review]
LGTM, but lets get one more green checkmark before merging.

File: pkg/services/auth/jwt.go
Comment (JamWils) [reply pkg/services/auth/jwt.go]: That way he doesn't have to mess with the internals of this function.  In my mind, Generate unit tests should focus on the internals.  By mocking it out from `server_test.go` he is able to just focus on testing `application.Server.Authenticate(...)`.  From there he is able to test both paths easily:

1. `Generate` returns a valid token
2. `Generate` returns an error

Without the mock, how does he test the sad path?  He can try to force `Generate` into an error state, but that introduces a cascading effect of failures within the test suites that can get annoying very quickly.  If `Generate` evolves because we sign the JWT with a certificate or encrypt information, then the tests in `server_test.go` will most likely fail and need to be updated as well because they are dependent on the internal implementation of `Genereate`.  

It's a trade off, disconnecting the two means as an engineeer updates `Generate` the engineer can focus on making sure that `jwt_test.go` has full coverage and the tests will clearly show which units have broken.  `server_test.go` is still good because ultimately it only cares with a valid token or error is returned.  End-to-end tests can test the full flow.  It's a trade off but it simplifies the code and makes it easier to test in my opinion.

File: pkg/services/auth/jwt.go
Comment (bigkevmcd) [inline pkg/services/auth/jwt.go:41]: I think this should be `internalJWTClient` ?

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: pkg/services/auth/jwt.go
Comment (bigkevmcd) [inline pkg/services/auth/jwt.go:32]: This should be documented (it's exported).

File: pkg/services/auth/jwt.go
Comment (bigkevmcd) [inline pkg/services/auth/jwt.go:37]: Missing a doc comment.

File: pkg/services/auth/jwt.go
Comment (bigkevmcd) [inline pkg/services/auth/jwt.go:38]: I think instead of using a package global, this should accept the random string.

And change the instantiation to provide the secure string to use.

```go
cfg := ApplicationConfig{App: app.New(nil, nil, nil, k, nil), JwtClient: auth.NewJwtClient(<insert random string here>)}
```

File: pkg/server/server.go
Comment (bigkevmcd) [inline pkg/server/server.go:333]: I'm not sure there's a lot of value in the [`GitProviderName`](https://github.com/weaveworks/weave-gitops/blob/cf1654f34912bae653be91df69d1e3d855624a90/pkg/gitproviders/factory.go#L15) type here as you can see it complicates comparisons.

I'd probably switch back to a `string` which is what the Go stdlib does https://pkg.go.dev/net/http#pkg-constants

File: pkg/server/server.go
Comment (bigkevmcd) [inline pkg/server/server.go:335]: You don't need to call `.Error()` on the error here, because you're using `%s` https://pkg.go.dev/fmt

> 3. If the %v verb is used with the # flag (%#v) and the operand implements the GoStringer interface, that will be invoked.
> If the format (which is implicitly %v for Println etc.) is valid for a string (%s %q %v %x %X), the following two rules apply:
> 4. If an operand implements the error interface, the Error method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).

So, because `BadErrProvider` implements `.Error()` it will be converted to a string automatically.

File: pkg/server/server_test.go
Comment (bigkevmcd) [inline pkg/server/server_test.go:131]: Bonus points: I think I'd write a custom matcher for these https://onsi.github.io/gomega/#adding-your-own-matchers that would look like `Expect(err).To(MatchGRPCError(codes.InvalidArgument, ErrEmptyAccessToken))`

File: pkg/server/server_test.go
Comment (bigkevmcd) [inline pkg/server/server_test.go:250]: You could simplify this to...

```go
strings.NewReader(`{"accessToken":"sometoken"}`)
```

Which would remove the string quoting.

File: pkg/server/server_test.go
Comment (bigkevmcd) [inline pkg/server/server_test.go:254]: You could probably use [`MatchJSON`](https://onsi.github.io/gomega/#matchjsonjson-interface) here

**Review by bigkevmcd — COMMENTED** [review]
Some comments.

**Review by luizbafilho — APPROVED** [review]

**Pull request merged**
